source/Timer.cpp

```diff
@@ -0,0 +1,271 @@
+/* Timer.cpp
+Copyright (c) 2023 by Timothy Collett
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Timer.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "GameData.h"
+#include "Logger.h"
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+#include "Ship.h"
+#include "StellarObject.h"
+#include "UI.h"
+
+using namespace std;
+
+Timer::Timer(const DataNode &node, Mission *mission)
+{
+	Load(node, mission);
+}
+
+void Timer::Load(const DataNode &node, Mission *mission)
+{
+	if(node.Size() > 1)
+		name = node.Token(1);
+	if(node.Size() > 2)
+		base = static_cast<int64_t>(node.Value(2));
+	if(node.Size() > 3)
+		rand = static_cast<uint32_t>(node.Value(3));
+
+	this->mission = mission;
+
+	timeToWait = base + Random::Int(rand);
+
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "idle")
+			requireIdle = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "proximity")
+		{
+			if(child.Size() > 1)
+				proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
+			if(child.HasChildren())
+			{
+				const DataNode &firstGrand = (*child.begin());
+				proximityCenter = GameData::Planets().Get(firstGrand.Token(0));
+			}
+		}
+		else if(child.Token(0) == "uncloaked")
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset")
+		{
+			if(child.Size() > 1 && child.Token(1) == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(child.Size() > 1 && child.Token(1) == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(child.Size() > 1 && child.Token(1) == "pause")
+				resetCondition = ResetCondition::PAUSE;
```

Comment:
I think you can simplify this a bit. Are there any cases for timers to require say uncloaked and being in a zone but only resetting the timer if the ship leaves the zone? I would expect like most if not all timers to apply any reset if any of their conditions are no longer met.

Reply:
That would be `reset pause`—I can make that the default, and add an explicit `reset never` option instead, if you think that makes more sense.

I don't know what all the use cases would be; I believe that when I wrote this, there had been a Discord conversation a day or two previously that had given some ideas (such as the "uncloaked" part, and having to wait near a particular spöb), so I was just trying to make sure it would _at least_ accommodate those.